File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.java
class Solution {
    int[][] dirs = new int[][]{{1, 0}, {1, -1}, {0, -1}, {-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}};
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        //corner
        if(grid[0][0] == 1 || grid[n - 1][n - 1] == 1) return -1;
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[n][n];
        int[] start = new int[]{0, 0};
        int[] end = new int[]{n -1, n - 1};
        
        q.offer(start);
        visited[start[0]][start[1]] = true;
        int step  = 1;
        while(!q.isEmpty()) {
            int size = q.size();
            for(int i = 0; i < size; i++) {
                int[] node = q.poll();
                if(node[0] == end[0] && node[1] == end[1]) return step;
                for(int[] dir : dirs) {
                    int x = node[0] + dir[0];
                    int y = node[1] + dir[1];
                    if(x < 0 || x >= n || y < 0 || y >= n || visited[x][y] || grid[x][y] == 1) continue;
                    q.offer(new int[]{x, y});
                    visited[x][y] = true;
                }
            }
            step++;
        }
        return -1;
    }
}

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        n = len(grid)
        if grid[0][0] == 1 or grid[n- 1][n - 1] == 1:
            return -1
        q=deque([(0, 0)])
        seen = set()
        seen.add((0, 0))
        dis = 1
        while q:
            size = len(q)
            for i in range(size):
                r, c = q.popleft()
                if r == n - 1 and c == n - 1:
                    return dis
                for x, y in (1, 0), (1, -1),(0, -1), (-1, -1), (-1, 0), (-1, 1), (0, 1), (1, 1):        
                    nr, nc = r + x, c + y
                    if nr >= 0 and nr < n and nc >= 0 and nc < n and grid[nr][nc]== 0 and  (nr, nc) not in seen:
                        q.append((nr, nc))
                        seen.add((nr, nc))
            dis += 1
        return -1
        